File: unijava/src/org/unified/uni/Token_reader.java
package org.unified.uni;

/**
 * Uni parser
 * <p>
 * Created by Michael Vlasov on 19.04.2018.
 * <p>
 * Copyright (c) 2017 Michael Vlasov
 */
public class Token_reader {
	public static class Error extends Exception {
		public String message;


		public Error(String message) {
			super();
			this.message = message;
		}


		@Override
		public String toString() {
			return message;
		}
	}


	public interface Tokenizer {
		Token apply(Char_reader input) throws Error;
	}


	public Token_reader(String source, Tokenizer tokenizer) throws Error {
		this.tokenizer = tokenizer;
		input = new Char_reader(source);
		has_current = true;
		passed = null;
		current_pos = input.current_pos;
		current = null;
		next();
	}


	public String current_context() {
		return input.context_at(current_pos);
	}


	public boolean pass(Token token) throws Error {
		if (!has_current || current.type != token.type) {
			return false;
		}
		passed = current;
		next();
		return true;
	}


	public void pass_required(Token token) throws Error {
		if (!pass(token)) {
			throw new Error("Expected " + Token.names.get(token.type) + " at: " + current_context());
		}
	}


	public void next() throws Error {
		if (!has_current) {
			return;
		}
		if (!input.has_current) {
			current = null;
			has_current = false;
			return;
		}
		current_pos = input.current_pos;
		current = tokenizer.apply(this.input);
		if (current == null) {
			has_current = false;
			if (input.has_current) {
				throw new Error("Error at " + current_context());
			}
		}
	}


	private final Char_reader input;
	boolean has_current;
	Token passed;
	private int current_pos;
	Token current;
	private final Tokenizer tokenizer;
}

File: uniswift/Uni/Uni/Lexer/TokenReader.swift
//
//  TokenReader.swift
//  Uni
//
//  Created by Michael Vlasov on 06.05.2018.
//  Copyright © 2018 Unified. All rights reserved.
//

import Foundation

public enum ParsingError: Error, CustomStringConvertible, CustomDebugStringConvertible {
	case expected(required: LexicalToken, context: String)
	case unexpected(token: LexicalToken, context: String)
	case error(message: String, context: String)
	
	public var localizedDescription: String {
		switch self {
		case .expected(let required, let context):
			return "Expected \(required.nameString) at: \(context)"
		case .unexpected(let token, let context):
			return "Unexpected \(token.nameString) at: \(context)"
		case .error(let message, let context):
			return "\(message) at \(context)"
		}
	}
	
	public var description: String {
		return localizedDescription
	}
	
	public var debugDescription: String {
		return localizedDescription
	}
}


public class TokenReader {
	
	public private(set) var hasCurrent: Bool
	public private(set) var passed = LexicalToken.empty
	public private(set) var current = LexicalToken.empty

	
	public init(chars: CharReader, tokenizer: @escaping (CharReader) throws -> LexicalToken?) throws {
		self.tokenizer = tokenizer
		self.chars = chars
		currentPos = chars.currentPos
		hasCurrent = true
		passed = .empty
		current = .empty
		try next()
	}
	
	
	public var currentContext: String {
		return chars.contextAt(pos: currentPos)
	}
	
	
	public func pass(token: LexicalToken) throws -> Bool {
		guard hasCurrent && current.name == token.name else {
			return false
		}
		passed = current
		try next()
		return true
	}
	
	
	public func pass(required: LexicalToken) throws {
		guard try pass(token: required) else {
			throw ParsingError.expected(required: required, context: currentContext)
		}
	}
	
	
	public func next() throws {
		guard hasCurrent else {
			return
		}
		guard chars.hasCurrent else {
			current = .empty
			hasCurrent = false
			return
		}
		currentPos = chars.currentPos
		guard let token = try tokenizer(chars) else {
			hasCurrent = false
			if chars.hasCurrent {
				throw ParsingError.unexpected(token: current, context: currentContext)
			}
			return
		}
		current = token
	}
	
	// MARK: - Internals
	
	private let chars: CharReader
	private var currentPos: String.UnicodeScalarIndex
	private let tokenizer: (CharReader) throws -> LexicalToken?
}

